### ISSUE TITLE
Metric learning on a dataset with multi-modal classes sharing some modalities

### ISSUE THREAD

**Issue by dimitry12** [KevinMusgrave/pytorch-metric-learning#273]
I have a dataset of images, with each class corresponding to "author of the image". In other words, it's a labelled dataset of images, and each image is labelled with single user ID. Dataset contains tens-to-hundreds images per author/class, and a few thousand authors/classes.

After reviewing the dataset I noticed a pattern:

- Classes in this dataset are multimodal in a sense that each user usually has several distinct "interests" making them produce an image. In other words, images authored by a given user form several "clusters".
- Yet, some "interests" are shared across users. For example, there are many users all having "cat" images or "cute babies" images.

"Interest" here is a "latent" and is not observed (other than me concluding its existence from starring at the dataset).

My goal is: given a new unseen author (represented as a set of her images), decide how likely a given new unseen image is to belong to this author.

I've tried straightforward metric-learning on the images achieving `P@1=0.35`, `P@R=0.11`, `MAP@R=0.06` on 1000 unseen authors/classes (each with a couple hundred images).

Given that model only finds embedding for a single image at a time, and given the structure of the dataset, I imagine the following happens:

- all images from heavily shared modalities end up close together in the embeddings-space - either as one cluster or multiple;
- unique modalities are their own clusters in the embeddings-space.

I am looking for advice regarding:
- how to best confirm or disconfirm my intuitions above about the resulting structure of the embeddings space?
- alternative approaches to solving this: with or without metric learning

**Comment by KevinMusgrave** [comment]
I don't have any experience with this, but here are some ideas anyway.

### Checking your intuition
To get the "content" label of the images, you could use a network pretrained on Imagenet or some other large dataset, and get its predictions for all the images in your dataset. Then you can plot the embedding space with the predicted content labels, to see if you are correct about the shared modalities.

### Solving the actual problem
During inference, you could filter the embeddings by user, and then see for which user the new image is not an outlier. In pseudocode:
```python
for user in users:
    curr_embeddings = all_embeddings[idx_of_imgs_by_user]
    if not outlier(curr_embeddings, new_embedding):
        return user
```
You would need criteria for determining if an embedding is an outlier.

Here's a paper that might be related: https://openaccess.thecvf.com/content_cvpr_2016/papers/Thomas_Seeing_Behind_the_CVPR_2016_paper.pdf
